File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
function solution(dirs) {
    let x = 0;
    let y = 0;
    const visited = new Set();
    
    for(let i=0; i<dirs.length; i+=1){
        const dir = dirs[i];
        const [nx, ny] = move(x,y,dir);
        
        if(!isMoveable(nx,ny)){
            continue;
        }
        
        visited.add(`${x}${y}${nx}${ny}`)
        visited.add(`${nx}${ny}${x}${y}`)
        
        x = nx;
        y = ny
    }
    
    return visited.size / 2
}

function move(x,y,dir){
    if(dir==='U'){
        return [x,y+1]
    }
    if(dir==='D'){
        return [x,y-1]
    }
    if(dir==='L'){
        return [x-1,y]
    }
    if(dir==='R'){
        return [x+1,y]
    }
}

function isMoveable(x,y){
    return x>=-5 && x<=5 && y<=5 && y>=-5
}

File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
import copy

def solution(dirs):
    cur = [0,0]
    dir = {'L': [-1, 0], 'R': [1, 0], 'U': [0, 1], 'D': [0, -1]}
    result = []
    for d in dirs:
        if(possible(cur, dir[d])):
            tmp = copy.deepcopy(cur)
            cur[0] = cur[0] + dir[d][0]
            cur[1] = cur[1] + dir[d][1]
            result.append(f'{",".join(map(str,tmp))} {",".join(map(str,cur))}')
            result.append(f'{",".join(map(str,cur))} {",".join(map(str,tmp))}')
    return (len(set(result))/2)
        
def possible(cur, next):
    if (cur[0]+next[0]<-5 or cur[0]+next[0]>5 or cur[1]+next[1]<-5 or cur[1]+next[1]>5): return False
    return True